futures-util/src/io/write_all.rs

```diff
@@ -40,3 +41,239 @@ impl<W: AsyncWrite + ?Sized + Unpin> Future for WriteAll<'_, W> {
         Poll::Ready(Ok(()))
     }
 }
+
+/// Future for the
+/// [`write_all_vectored`](super::AsyncWriteExt::write_all_vectored) method.
+#[derive(Debug)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub struct WriteAllVectored<'a, W: ?Sized + Unpin> {
+    writer: &'a mut W,
+    bufs: &'a mut [IoSlice<'a>],
+}
+
+impl<W: ?Sized + Unpin> Unpin for WriteAllVectored<'_, W> {}
+
+impl<'a, W: AsyncWrite + ?Sized + Unpin> WriteAllVectored<'a, W> {
+    pub(super) fn new(writer: &'a mut W, bufs: &'a mut [IoSlice<'a>]) -> Self {
+        WriteAllVectored { writer, bufs }
+    }
+}
+
+impl<W: AsyncWrite + ?Sized + Unpin> Future for WriteAllVectored<'_, W> {
+    type Output = io::Result<()>;
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<io::Result<()>> {
+        let this = &mut *self;
+        while !this.bufs.is_empty() {
+            let n =
+                ready!(Pin::new(&mut this.writer)
+                    .poll_write_vectored(cx, this.bufs))?;
+            if n == 0 {
+                return Poll::Ready(Err(io::ErrorKind::WriteZero.into()));
+            } else {
+                let mut remove = 0;
+                let mut accumulated_len = 0;
+                for buf in this.bufs.iter() {
+                    if accumulated_len + buf.len() > n {
+                        break;
+                    } else {
+                        accumulated_len += buf.len();
+                        remove += 1;
+                    }
+                }
+
+                this.bufs = mem::replace(&mut this.bufs, &mut [])
+                    .split_at_mut(remove)
+                    .1;
+
+                // Bytes already written from the (now) first buffer.
+                let bytes_written = n - accumulated_len;
+                if bytes_written != 0 {
+                    let bytes = unsafe {
+                        // We create a new slice unsafely to get around lifetime
+                        // issues. Since all buffers need to life at least as long
+                        // as WriteAllVectored, by definition of lifetime 'a, this
+                        // should be safe.
+                        slice::from_raw_parts(
+                            this.bufs[0].as_ptr().add(bytes_written),
+                            this.bufs[0].len() - bytes_written,
+                        )
+                    };
```

Comment:
This seems to indicate a missing `IoSlice` API; either something like `into_slice` so that you could get the slice with correct lifetime back and use `split_at`, or having `split_at` directly on `IoSlice`. Would be good to open an issue about adding something to move this unsafety to `std`.

Reply:
Or better yet and `advance` method that would do this for use.

Reply:
Created https://github.com/rust-lang/rust/issues/62726.

Reply:
This is now available using the `io_slice_advance` Rust feature, tracking issue: https://github.com/rust-lang/rust/issues/62726.